include/nbl/asset/utils/CSPIRVIntrospector.h

```diff
@@ -13,172 +17,657 @@
+				core::vector<SDescriptorVarInfo<>> m_descriptorSetBindings[DESCRIPTOR_SET_COUNT];
+				// The idea is that we construct with based_span (so we can add `.data()` when accessing)
+				// then just convert from `SMemoryBlock<true>` to `SMemoryBlock<false>`
+				// in-place when filling in the `descriptorSetBinding` vector which we pass to ctor
+				core::vector<char> m_memPool;
+		};
+		class CPipelineIntrospectionData final : public CIntrospectionData
 		{
-			std::string entryPoint;
-			core::smart_refctd_ptr<const ICPUShader> cpuShader;
+			public:
+				struct SDescriptorInfo final : CIntrospectionData::SDescriptorInfo
+				{
+					inline bool isArray() const {return stride;}
+					inline bool isRuntimeSized() const {return isArray() && count==0;}
+
+					uint32_t count : 21 = 0;
+					uint32_t stride : 11 = 0;
+					// Which shader stages touch it
+					core::bitflag<ICPUShader::E_SHADER_STAGE> stageMask = ICPUShader::ESS_UNKNOWN;
+				};
+				//
+				inline CPipelineIntrospectionData()
+				{
+					std::fill(m_pushConstantBytes.begin(),m_pushConstantBytes.end(),ICPUShader::ESS_UNKNOWN);
+					std::fill(m_highestBindingNumbers.begin(), m_highestBindingNumbers.end(), -1);
+				}
 
-			bool operator==(const SIntrospectionParams& rhs) const
-			{
-				if (entryPoint != rhs.entryPoint)
-					return false;
-				if (!rhs.cpuShader)
-					return false;
-				if (cpuShader->getStage() != rhs.cpuShader->getStage())
-					return false;
-				if (cpuShader->getContentType() != rhs.cpuShader->getContentType())
-					return false;
-				if (cpuShader->getContent()->getSize() != rhs.cpuShader->getContent()->getSize())
-					return false;
-				return memcmp(cpuShader->getContent()->getPointer(), rhs.cpuShader->getContent()->getPointer(), cpuShader->getContent()->getSize()) == 0;
-			}
+				// returns true if successfully added all the info to self, false if incompatible with what's already in our pipeline or incomplete (e.g. missing spec constants)
+				NBL_API2 bool merge(const CStageIntrospectionData* stageData, const ICPUShader::SSpecInfoBase::spec_constant_map_t* specConstants=nullptr);
+
+				//
+				NBL_API2 core::smart_refctd_dynamic_array<SPushConstantRange> createPushConstantRangesFromIntrospection(core::smart_refctd_ptr<const CStageIntrospectionData>& introspection);
+				NBL_API2 core::smart_refctd_ptr<ICPUDescriptorSetLayout> createApproximateDescriptorSetLayoutFromIntrospection(const uint32_t setID);
+				NBL_API2 core::smart_refctd_ptr<ICPUPipelineLayout> createApproximatePipelineLayoutFromIntrospection(core::smart_refctd_ptr<const CStageIntrospectionData>& introspection);
+
+			protected:
+				// ESS_UNKNOWN on a byte means its not declared in any shader merged so far
+				std::array<core::bitflag<ICPUShader::E_SHADER_STAGE>,MaxPushConstantsSize> m_pushConstantBytes;
+				//
+				struct Hash
+				{
+					inline size_t operator()(const SDescriptorInfo& item) const {return item.binding;}
```

Comment:
you need to put `item.binding` through `std::hash`, its necessary for the hashmap to throw the buckets around nicely